trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    /// Convert a [`GQLType`] to a [`Type`].
+    pub(crate) fn from_type(ty: &GQLType) -> Type {
+        let mut base = &ty.base;
+
+        let mut mask = if ty.nullable { 0 } else { Modifiers::NON_NULLABLE_MASK };
+
+        let mut i = 0;
+
+        while let BaseType::List(ty_inside_list) = base {
+            mask |= Modifiers::LIST_MASK << i;
+            i += 2;
+            if i > Modifiers::MAX_LIST_DEPTH * 2 {
+                panic!("too many nested lists");
+            }
+            if !ty_inside_list.nullable {
+                mask |= Modifiers::NON_NULLABLE_MASK << i;
+            }
+            base = &ty_inside_list.base;
+        }
+
+        let BaseType::Named(name) = base else {
+            unreachable!(
+                "should be impossible to get a non-named type after looping through all list types"
+            )
+        };
+        Type { base: name.to_string().into(), modifiers: Modifiers { mask } }
+    }
+}
+
+impl Display for Type {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // left
+        {
+            let mut current = Some(self.modifiers.clone());
+            while let Some(mods) = current {
+                if mods.is_list() {
+                    write!(f, "[")?;
+                }
+                current = mods.as_list();
+            }
+        }
+
+        write!(f, "{}", self.base)?;
+
+        let mut builder = String::new();
+
+        // right
+        {
+            let mut current = Some(self.modifiers.clone());
+            while let Some(mods) = current {
+                if !mods.nullable() {
+                    builder.push('!');
+                }
+                if mods.is_list() {
+                    builder.push(']');
+                }
+                current = mods.as_list();
+            }
+            write!(f, "{}", builder.chars().rev().collect::<String>())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Serialize for Type {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for Type {
+    fn deserialize<D>(deserializer: D) -> Result<Type, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct TypeDeserializer;
+
+        impl<'de> Visitor<'de> for TypeDeserializer {
+            type Value = Type;
+
+            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
+                formatter.write_str("GraphQL type")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let ty = Type::new(s)
+                    .ok_or_else(|| serde::de::Error::custom("not a valid GraphQL type"))?;
+                Ok(ty)
+            }
+        }
+
+        deserializer.deserialize_str(TypeDeserializer)
+    }
+}
+
+#[cfg(test)]
```

Comment:
This needs lots more tests. A few examples off the top of my head:
- Complex types with mixed nullability across multiple nested lists get parsed, represented, and `.to_string()`-ed correctly.
- More in-depth tests for `.to_string()` in general.
- Tests for the methods that allow building up types by wrapping existing types in a list and/or setting nullability.